### ISSUE TITLE
Make public key length available via the API

### ISSUE THREAD

**Issue by mbartosch** [gknocke/Crypt-PKCS10#10]
It would be nice to have an accessor for obtaining the public key length. This probably requires postprocessing of the internal _pubkey variable, as the key length itself is not part of the ASN.1 modulus.

**Comment by tlhackque** [comment]
Yes, I noted that in #9.

The key length **is** available as csr->subjectPublicKey * 4.  (Yes, it's ugly...and somewhat fragile, though key lengths have all been multiples of 4.)

I agree that a mechanism that explicitly returns the length and is consistent with EC would be better.

Perhaps 
````
csr->subjectPublicKeyParams => 
{
  KeyType => oid,      # caller can use oid2name can return the name
  KeyLength => bits, # for non-rsa, may require a table of OIDs (e.g. keys with odd lengths)
}
````
This would allow for future key types that might have additional parameters.

If you develop one, remember to add it to the pod as well as the tests...

The mechanics of a release are a bit involved, so please note whether more work is expected in your pull request(s).

**Comment by mbartosch** [comment]
Thanks for the quick release, in fact this was very appreciated as we intend to move OpenXPKI to Crypt::PKCS10 soon.
Parsing the parameters and key length is secondary. I will probably work on this later. After a pull request I will indicate whether more is to be expected within reasonable time.

**Comment by tlhackque** [comment]
It turns out that you **can't** extract the key length in all cases.  

After parsing the key into its components, one gets the asn.1 representation of the modulus - which is an INTEGER.  In asn.1, the INTEGER encoding minimizes space by eliminating leading all-zero bytes (octets).  So if you happen to get a key with 8 leading zeroes, there's no way to tell.  That's unlikely in any given case, but it **will** happen.  On the other hand, asn.1 stores INTEGERs in bytes, so it may introduce up to 7 bits of MSB padding.  (There's a minor complication in that the modulus is unsigned, but asn.1 is always signed 2s complement.  So half the moduli have an extra byte tacked on in the ASN.1 representation.  That's easy to deal with, as it's the only case where INTEGER DER will have a leading byte of zero.)

I implemented an experimental version of _subjectPublicKeyParams()_, which will return an estimate of the key size as well as some other useful information from the public key and its parameters.   As a practical matter, I doubt the key length will be observed to be short by more than a few bits; but consumers need to account for the fuzzy value.

For now, this version is in my repo (https://github.com/tlhackque/Crypt-PKCS10).  Consider it volatile.

**Feedback welcome**.  It would be particularly entertaining to see cases where real software generates keys whose reported key length is unexpected.  At least one should be added to the tests.

I'll push this to the project repo and release when I have sufficient feedback.

BTW, _pubkey is just the PEM version of the key before PCS10 decodes the asn.1- not very useful in Perl code.  I keep it around for output via the _subjectPublicKey(1)_ method.  

You could run it through an external tool, e.g. OpenSSL.  But **as far as I can tell, the inaccuracy is inherent in asn.1.**  So I don't think it's possible to do better.  One could round up - but that assumes knowledge of probable key lengths.  I don't want to do that: it may overstate the key length.  A caller may choose to do things like that based on its knowledge of how the value will be used.

The OpenSSL folks agree:  see the NOTES section of https://www.openssl.org/docs/man1.1.0/crypto/BN_num_bits.html

**Comment by mbartosch** [comment]
Thank you so much for investing time on this topic. 
I agree with you, it is quite difficult to determine the correct key size by observing the DER representation of the encoded public key.
One thing I noticed in the past is that most implementations seem to make sure that the MSB of the public key is 1 (e. g. by discarding generated keys for which this is not true), assuring that the maximum number of bits are allocated within the ASN.1 representation. I found that this is the case for many but not all crypto toolkits. Some negative examples are earlier versions of GSKit (IBM).
In any case, returning the exact bit length of the public key is totally sufficient in my opinion. In the OpenXPKI project we have been dealing with the problem of statistical variations of the measured public key length for some time and have found acceptable workarounds for the problem (e. g. ranges of bit length).

**Comment by tlhackque** [comment]
As usual, it took more time to explain than to write the code.

I don't find the highest bit set (which is what OpenSSL does).  I use a different approximation that's simpler because of the Convert::ASN1 representation of the integers.

The parsed value of the key is a Math::BigInt, which I convert to hex, then count the number of digits.  

Multiplying by 4 (bits/nibble) approximates the length. (EC keys are handled slightly differently, as they're supposed to be uncompressed.  Since both x and y are present, I just multiply by 2 instead of 4.  If compressed, which the RFCs discourage, the key may appear slightly longer depending on how the quadrant bit is represented.)

Thus, if there's a '1' in any of the 4 MSB (or maybe it's 8 - I didn't check), the length will round up to the expected value.   At least for common key sizes.  Anyhow, it's easier than doing logical operations on a BigInt, and the inaccuracy will be the same (or usually less).  For random keys, only 1/16 will have all 4 MSB zero, or 1/256 all 8.  Finding the exact highest bit set will be wrong 1/2 the time.  You still need to deal with fuzz, but you'll notice it less.  I think it's close enough.

The problem with discarding keys generated with the MSB == 0 is that the effective key size is reduced by 1 bit.  A case of cosmetics winning over function.  An attacker is guaranteed that the LSB is set, and can assume that the MSB is also set... 

We live in an imperfect world.

**Comment by tlhackque** [comment]
Fixed in V1.7

EC key length now comes from the OID, courtesy of Crypt::PK::ECC.
The others are as described above.
